corehq/apps/domain/views/settings.py

```diff
@@ -517,3 +520,87 @@ class ManageDomainMobileWorkersView(ManageMobileWorkersMixin, BaseAdminProjectSe
     page_title = gettext_lazy("Manage Mobile Workers")
     template_name = 'enterprise/manage_mobile_workers.html'
     urlname = 'domain_manage_mobile_workers'
+
+
+@method_decorator(toggles.CUSTOM_DOMAIN_BANNER_ALERTS.required_decorator(), name='dispatch')
+class ManageDomainAlertsView(BaseAdminProjectSettingsView):
+    template_name = 'domain/admin/manage_alerts.html'
+    urlname = 'domain_manage_alerts'
+    page_title = gettext_lazy("Manage Project Alerts")
+
+    @property
+    def page_context(self):
+        return {
+            'form': self.form,
+            'alerts': [
+                {
+                    'active': alert.active,
+                    'html': alert.html,
+                    'id': alert.id,
+                    'created_by_user': alert.created_by_user,
+                }
+                for alert in CommCareHQAlert.objects.filter(created_by_domain=self.domain)
+            ]
+        }
+
+    @cached_property
+    def form(self):
+        if self.request.method == 'POST':
+            return DomainAlertForm(self.request.POST)
```

Comment:
For my own curiosity: the `text` field is the only thing that `DomainAlertForm` cares about, right? Does it matter if there are other fields present in `request.POST`?

Reply:
Ya, that is the only field for now.
Unless someone adds something manually by modifying the HTML, `text` should be the only field.

Alert's do support a start/end time too which could be added in future if needed.

Reply:
> Unless someone adds something manually by modifying the HTML, `text` should be the only field.

What about

https://github.com/dimagi/commcare-hq/blob/49c42c2180b1b88b6c0aff7f341cfe48c02f3bc7/corehq/apps/domain/views/settings.py#L553

and

https://github.com/dimagi/commcare-hq/blob/49c42c2180b1b88b6c0aff7f341cfe48c02f3bc7/corehq/apps/domain/views/settings.py#L565

Reply:
Does that mean it is dangerous to reference `self.form` in this method as well as some methods that it calls (e.g., `_apply_command`) except in very particular circumstances? That was not obvious to me, and seems like an easy thing for future readers to overlook.

I suggest refactoring the code to make that not possible or at least much harder to do by mistake.

Reply:
Not sure I understand. If someone calls `self.form` when it should not be, it would just give back a form with no data in it.

The entry function `post`, supports different operations (create or update/delete) and hence can get requests from different forms on the page. It decides which operations to do based on data passed in the request.

In order to make it more clear, is this better?
```
def create_alert_form()
    if self.request.method == 'POST':
        return DomainAlertForm(self.request.POST)
    return DomainAlertForm()

def post(self, request, *args, **kwargs):
    command = request.POST.get('command')
    if command:
        self._apply_command(request, command)
    else:
        if self.create_alert_form.is_valid():
            self._create_alert()
            messages.success(request, _("Alert saved!"))
        else:
            messages.error(request, _("There was an error saving your alert. Please try again!"))
            return self.get(request, *args, **kwargs)
    return HttpResponseRedirect(self.page_url)

def _apply_command(self, request, command):
    # ...
```
or you are proposing a structural change to the way the view and forms are interacting?
Like to just create the alert on the page via POST but handle activation/deactivation/deletion via a separate URL?
Or That a view should have only one form and that all operations should ideally be done by the same form?

Reply:
I was thinking of something more like this
```py
def post(self, request, *args, **kwargs):
    command = request.POST.get('command')
    if command:
        self._apply_command(request, command)
    else:
        form = DomainAlertForm(request.POST)
        if form.is_valid():
            self._create_alert(form)
            ...
```
Where `form` is not a property of `ManageDomainAlertsView`. That way it cannot be referenced in an invalid context.

The reference in `page_context` still needs to be handled. It is unclear whether that method is ever called during a POST request that is not a `DomainAlertForm` submission, so unclear whether it was safe to reference `self.form` in there.

Reply:
When there is an error in the form, the state of the form is to be retained so that the errors are displayed. A common way to do that is to have the form memoized, so when we return the `get` view, which will then call `page_context` and same form will be used.
So, until that is changed, `form` needs to be a property of `ManageDomainAlertsView`

Reply:
To unblock this PR I'm going to drop this request for change.

My main complaint is that the context expected by the `form` property is dependent on things that are defined relatively distantly from the property. For example, it may not be obvious to a future developer who wants to change the `post` method that their change may cause `page_context` to access `self.form` when the submitted form does not match the fields of `DomainAlertForm`.

Said a slightly different way, it may not be obvious that the `post` method accepts multiple "different forms." This was not obvious to me.

Reply:
Thanks for the relaxation @millerdev 

> it may not be obvious that the post method accepts multiple "different forms." This was not obvious to me.

This, however stuck with me and tried to make this better in the following commit from how I proposed an alternate earlier.
https://github.com/dimagi/commcare-hq/pull/33592/commits/a6808a8c9af4f974b8963880524d15737f8e0a06

The diff might not be easier to review so i left some comments. There is no functional feature change in that commit.
Please let me know if that makes the code clearer. Thanks for your patience with this.

Reply:
That change did simplify `ManageDomainAlertsView.post`, but with the cost of adding two new URL endpoints. It's probably a nitpick  at this point, but I think it would have been cleaner to keep a single URL endpoint and use a different technique ensure that the logic in `ManageDomainAlertsView.form` could not be mistakenly called with the wrong form. However, I'm not sure how to better suggest a different approach apart from suggestions above. This solution is good enough.